### PR TITLE
Custom operator precedence

### LINKED ISSUE #1007: Precedence of custom operators?
My main intention here is to file a documentation bug; there doesn't seem to be any discussion of the precedence of custom operators in the Reference.

My concern here is that to deal with #1006 I am currently doing

`operator at<T>(A: T[], i: number) A.at(i) as T`

but this `at` has the wrong precedence: `[1,10,5] at 2-1` is 4, not 10 as it "should" be. Anything I can do? I.e., a secondary part of this issue would be ability to influence the precedence, at least among a few options, like maybe default (=medium), tight, and loose.

### PR DESCRIPTION
**Pull request by edemaine**
I'm not sure this syntax is perfect, and happy to have more back-and-forth here. But the cool thing is that the precedence mechanism works!

Fixes #1007

### CODE REVIEW

File: test/binary-op.civet [hunk 1]
```diff
@@ -827,3 +827,79 @@ describe "custom identifier infix operators", ->
...
+      same as symbol
+      ---
+      operator mul same (*) (a, b)
+        a*b
+      a+b*c mul d*e+f
+      ---
+      function mul (a, b) {
+        return a*b
+      }
+      a+mul(b*c, d)*e+f
```
Comment (STRd6) [inline test/binary-op.civet:842]: If `mul` is the same precedence as `*` shouldn't it be:

`a+b*mul(c, d)*e+f`

Comment (edemaine) [reply test/binary-op.civet:842]: No, it's acting left-associative with `*`, so `b*c` goes first.

**Comment by bbrk24** [comment]
Is there a way to specify associativity, or are custom operators always left-associative?

**Review by STRd6 — APPROVED** [review]
Happy to get this out experimentally while we can think about nicer syntax.

**Comment by edemaine** [comment]
@bbrk24 Currently the engine only supports left associativity, so that's all for now. It should be easy to add right associativity (and invalid associativity) in the future.

**Pull request merged**